### ISSUE TITLE
[java] DoNotTerminateVM：false negative when system exit in lambda

### ISSUE THREAD

**Issue by mohui1999** [pmd/pmd#4469]
**Affects PMD Version:** 7.0.0

**Rule:** [DoNotTerminateVM](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#donotterminatevm)

**Description:**
The DoNotTerminateVM rule checks that web applications do not call System.exit() or its equivalents. However, it may not report a violation when `System.exit()` is called from a lambda expression, which may result in a false negative.

**Code Sample demonstrating the issue:**
Maybe a false negative / positive:

```java
public class SystemCall {
    public static void main(String[] args) {
       
        Runnable runnable = () -> {
            // NEVER DO THIS IN AN APP SERVER !!!
            System.exit(0);
        };

        new Thread(runnable).start();
    }
}
```

Correctly reported code:

```java
public class SystemCall {
    public static void main(String[] args) {
        new Runnable() {
            public void run() {
                // NEVER DO THIS IN A APP SERVER !!!
                System.exit(0);
            }
        };
    }
}
```

**Expected outcome:**
`java:6:   DoNotTerminateVM:       System.exit() should not be used in J2EE/JEE apps`
PMD should report a violation at line 6, but doesn't. This is a false negative.

**Running PMD through:** *[CLI]*

**Comment by oowekyala** [comment]
The reason it's ignored is not the lambda, but because this is in a main method. Whitelisting main methods is a practical compromise, as main methods should arguably use System exit somehow. They also are irrelevant to webapps AFAIA.

I would actually say your example with the anonymous class is a false positive.

**Comment by mohui1999** [comment]
Regarding the example with the anonymous class,  it is from the test case of the rule DoNotTerminateVM. However, since the expected-problems number is 1, it should report this rule in line 6. Should the expected-problems number change to 0?

```xml
    <test-code>
        <description>system exit in anonymous class inside main</description>
        <expected-problems>1</expected-problems>
        <expected-linenumbers>6</expected-linenumbers>
        <code><![CDATA[
public class SystemCall {
    public static void main(String[] args) {
        new Runnable() {
            public void run() {
                // NEVER DO THIS IN A APP SERVER !!!
                System.exit(0);
            }
        };
    }
}
        ]]></code>
    </test-code>
```

**Comment by oowekyala** [comment]
If we agree this is an FP, then yes this test case should be changed.

I would say one principle should be that `new Runnable() { ... System.exit() }`, `() -> System.exit(0);` or `System::exit` be treated the same. Then the question is whether they should all be reported or all allowed when found in a main method, and I would argue they should all be allowed.
